File: Data Structures & Algorithms/find-k-closest-elements/submission-1.java
class Solution {
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        List<Integer> res = new ArrayList<>();

        int left = 0 , right = arr.length - k;

        while (left < right){
            int mid = left + (right - left) / 2;

            if (x - arr[mid] > arr[mid + k] - x){
                left = mid + 1;
            }else {
                right = mid;
            }
        }

        for (int i = left; i < left + k; i ++){
            res.add(arr[i]);
        }

        return res;
        // for(int i = 0; i < arr.length; i++){
        //     for (int j = i + 1 ; j < arr.length; j++){
        //         int ax = Math.abs(arr[j] - x);
        //         int bx = Math.abs(arr[i] - x);
        //         if ( (ax < bx) || ((ax == bx) && (arr[j] < arr[i]))){
        //             int temp = arr[i];
        //             arr[i] = arr[j];
        //             arr[j] = temp;
        //         }
        //     }
        // }

        // List<Integer> res = new ArrayList<>();

        // for (int i = 0; i < k; i++){
        //     res.add(arr[i]);
        // }

        // Collections.sort(res);

        // return res;
    }
}

File: Data Structures & Algorithms/trapping-rain-water/submission-1.js
class Solution {
    /**
     * @param {number[]} height
     * @return {number}
     */
    trap(height) {
        let l = 0; 
        let r = height.length - 1;
        let res = 0;

        while (l < r){
            while(height[l] < height[l+1] && l < r){
                l++;
            }

            while(height[r] < height[r - 1] && l < r) {
                r--;
            }

            if(height[l] <= height[r] && l < r){
                const temp = height[l];
                l++
                while(height[l] < temp && l < r){
                    res += temp - height[l];
                    l++;
                }
            }else if (height[l] > height[r] && l < r){
                const temp = height[r];
                r--;
                while(height[r] < temp && l < r){
                    res += temp - height[r];
                    r--;
                }
            }
        }

        return res
    }
}
